File: BC-W3/TaskCpp/bcw4/Basic/List/List.cpp
7#include "List.h"

List::List(int capacity, double multiplier){
    this->capacity = capacity;
    this->multiplier = multiplier;
    this->current = 0;
    
    if ( this->capacity < 0 ) {
        throw ZeroLenException();
    }
    
    int* temp = (int*)malloc(sizeof(int) * this->capacity);
    
    if ( temp == NULL ) {
        throw OutOfMemoryException();
    }
    
    this->array = temp;
}
List::~List(){
    free(this->array);
}

int List::size() const {
    return this->current;
}
int List::max_size() const {
    return this->capacity;
}

void List::erase(int index) {
    for ( int i = index; i < this->current; i++ ) {
        this->array[i] = this->array[i+1];
    }
    
    this->current -= 1;
}
void List::insert(int value, int index) {
    this->current += 1;
    
    if ( this->current > this->capacity ) {
        int newCapacity = this->capacity * this->multiplier;
        int* newArray = (int*)realloc(this->array, sizeof(int) * newCapacity);
        
        if ( newArray == NULL ) {
            throw OutOfMemoryException();
        }
        
        this->capacity = newCapacity;
        this->array = newArray;
    }
    
    for ( int i = this->current - 1; i >= index; i-- ) {
        this->array[i+1] = this->array[i];
    }
    
    this->array[index] = value;
}
int List::find(int value) const {
    for ( int index = 0; index < this->current; index++ ) {
        if ( array[index] == value ) {
            return index;
        }
    }
}
void List::push_back(int value) {
    int newCurrent = this->current + 1;
    
    if ( newCurrent > this->capacity ) {
        int newCapacity = this->capacity * this->multiplier;
        int* newArray = (int*)realloc(this->array, sizeof(int) * newCapacity);
        
        if ( newArray == NULL ) {
            throw OutOfMemoryException();
        }
        
        this->capacity = newCapacity;
        this->array = newArray;
    }
    this->array[this->current] = value;
    this->current = newCurrent;
}
int List::pop_back() {
    if ( this->current == 0 ) {
        throw ZeroLenException();
    }
    this->current -= 1;
    
    return this->array[this->current];
}

void List::sort() {
    for ( int i = 1; i <= this->current; i++ ) {
        int temp = this->array[i];
        int j = i;
        
        for ( int prev = j - 1; j > 0 && this->array[prev] > temp; j--, prev-- ) {
            this->array[j] = this->array[prev];
        }
        this->array[j] = temp;
    }
}

int List::operator[](int index) const {
    return this->array[index];
}
bool List::operator==(const List& other) const {
    int i = 0;
    
    for ( ; i < this->current && i < other.current; i++ ) {
        if ( this->array[i] != other.array[i] ) {
            return false;
        }
    }
    if ( i == this->current && i == other.current ) {
        return true;
    }
    return false;
}
bool List::operator!=(const List& other) const {
    return !operator==(other);
}


std::ostream& operator<<(std::ostream& out, const List& list) {
    if ( list.size() == 0 ) {
        throw ZeroLenException();
    }
    int last = list.size() - 1;
    
    for ( int i = 0; i < last; i++ ) {
        out << list[i] << " ";
    }
    out << list[last];
    
    return out;
}

File: BC-W3/Java/java/java/List/List.java
import java.util.ArrayList;

public class List {
    private int capacity;
    private int current;
    private ArrayList<Integer> list;
    
    public List(int capacity) throws ZeroLenException {
        this.capacity = capacity;
        this.current = 0;
        this.list = new ArrayList<>(this.capacity);
        
        if ( this.capacity < 0 ) {
            throw new ZeroLenException();
        }
        
    }
    
    public List() throws ZeroLenException {
        this(100);
    }
    
    public int size() {
        return current;
    }
    
    public int max_size() {
        return capacity;
    }
    
    public void erase(int index) throws IndexOutOfBoundsException {
        if ( index < 0 || index >= this.current ) {
            throw new IndexOutOfBoundsException();
        }
        for ( int i = index; i < this.current - 1; i++ ) {
            this.list.set(i, this.list.get(i+1));
        }
        
        this.current -= 1;
    }
    
    public void insert(int value, int index) throws IndexOutOfBoundsException {
        if ( index < 0 || index >= this.current ) {
            throw new IndexOutOfBoundsException();
        }
        
        this.list.add(index, value);
        this.current += 1;
    }
    
    public int find(int value) {
        for ( int index = 0; index < this.current; index++) {
            if ( list.get(index) == value ) {
                return index;
            }
        }
        return -1;
    }
    
    public void push_back(int value) {
        this.list.add(current, value);
        this.current += 1;
    }
    
    public int pop_back() throws ZeroLenException {
        if ( this.current == 0 ) {
            throw new ZeroLenException();
        }
        
        this.current -= 1;
        
        return this.list.get(current);
    }
    
    public void sort() {
        for ( int i = 1; i < this.current; i++ ) {
            int temp = this.list.get(i);
            int j = i;
            
            for ( int prev = j - 1; j > 0 && this.list.get(prev) > temp; j--, prev-- ) {
                this.list.set(j, this.list.get(prev));
            }
            this.list.set(j, temp);
        }
    }
    
    public int get(int index) throws IndexOutOfBoundsException {
        if ( index < 0 || index >= this.current ) {
            throw new IndexOutOfBoundsException();
        }
        return this.list.get(index);
    }
    
    public boolean equals(List other) throws IndexOutOfBoundsException {
        int i = 0;
        
        for ( ; i < this.current && i < other.current; i++ ) {
            if ( this.get(i) != other.get(i) ) {
                return false;
            }
        }
        if ( i == this.current && i == other.current ) {
            return true;
        }
        return false;
    }
    
    public String toString() {
        int last = this.size() - 1;
        String out = "";
        
        for( int i = 0; i < last; i++ ) {
            out += this.list.get(i) + " ";
        }
        out += this.list.get(last);
        
        return out;
    }
}